### PR TITLE
Convert to AMD

### PR DESCRIPTION
**Pull request by muffinresearch**
This is step one of the move to UMD. FIrst convert to AMD style with karma-require to run the tests.

TODOs

* [x] Fix test flakiness
* [x] Work out why travis doesn't like the es6-promise code under karma-require.

It's easier to see the wood for the trees with: https://github.com/mozilla/fxpay/pull/101/files?w=1

### CODE REVIEW

File: lib/fxpay/products.js [hunk 1]
```diff
@@ -145,7 +148,6 @@
   Product.prototype.validateReceipt = function() {
     var receiptMap = this.getReceiptMap();
     var product = this;
-    var receipts = fxpay.getattr('receipts');
```
Comment (kumar303) [inline lib/fxpay/products.js]: did you try an inline ``require()`` here? Sort of lame though.

File: tests/test-main.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  }
+}
+
+requirejs.config({
+  // Karma serves files from '/base'
+  baseUrl: '/base/lib/fxpay/',
+
+  paths: {
+    'helper': '../../tests/helper',
+    'es6-promise': '/base/lib/bower_components/es6-promise/promise',
```
Comment (kumar303) [inline tests/test-main.js]: It looks like the ``es6-promise`` lib will call ``define()`` when it exists so can't we load this dynamically? It seems cumbersome to have to load all libs like this.

**Comment by kumar303** [comment]
As a first pass, you could try replacing all inline calls to ``fxpay.getattr()`` with inline ``require()`` calls. Maybe that would at least get tests passing?

**Comment by kumar303** [comment]
looks like you're on the right path here

File: tests/test-main.js (hunk 2 above)
Comment (muffinresearch) [inline tests/test-main.js]: There's something else causing the travis build to barf (despite it working fine locally). I'll fix that first and then see about reducing `baseUrl` to `/base/lib` which should mitigate the need for the path for es6-promise.

File: Gruntfile.js [hunk 3]
```diff
@@ -72,7 +57,7 @@ module.exports = function(grunt) {
           sourceMap: false,
         },
         files: {
-          'build/lib/fxpay.debug.js': libFiles,
+          'build/lib/fxpay.debug.js': 'lib/**/*.js',
```
Comment (kumar303) [inline Gruntfile.js]: How does this get the es6-promise shim?

**Comment by kumar303** [comment]
Are you sure that these circular dependencies will work? They won't bite us when we try to build a minified file or something? I guess the magic exports solves it.

File: tests/helper.js [hunk 4]
```diff
@@ -8,10 +11,9 @@
   exports.apiProduct = {guid: 'server-guid', name: "Name from API",
                         logo_url: "img.jpg"};
 
-
   exports.setUp = function setUp() {
     exports.server = sinon.fakeServer.create();
-    exports.settings = fxpay.settings.configure({
+    exports.settings = settings.configure({
```
Comment (kumar303) [inline tests/helper.js]: I'm not really sure how useful this is. You could alternatively replace ``helper.settings`` with required settings in each test. But no biggie.

**Comment by kumar303** [comment]
Instead of internal fxpay modules requiring other modules like ``utils`` shouldn't that be ``fxpay/utils``? I'm thinking that we may need the ``fxpay`` prefix to co-exist with other top-level modules, no?

**Comment by kumar303** [comment]
r+wc

File: tests/helper.js (hunk 4 above)
Comment (muffinresearch) [inline tests/helper.js]: Hmm, yeah I removed the things that reference that. So it should be ok to remove it.

File: Gruntfile.js (hunk 3 above)
Comment (muffinresearch) [inline Gruntfile.js]: This file building part will need to be replaced as change doesn't update the built file yet.

**Comment by muffinresearch** [comment]
> Instead of internal fxpay modules requiring other modules like utils shouldn't that be fxpay/utils? I'm thinking that we may need the fxpay prefix to co-exist with other top-level modules, no?

If that's a problem for us internally we can rethink it. I don't envisage this affecting the built file as that should be a single module when done.

File: tests/helper.js (hunk 4 above)
Comment (muffinresearch) [inline tests/helper.js]: Actually looks like I might have stashed that part. I'll do that now.

**Comment by muffinresearch** [comment]
Interesting. After removing all the helper.settings revealed the same test behaviour I was seeing before:

1__bash

After some head-scratching I found there was a real problem, which was caused by not re-initialising the onsuccess handler. 

That's now fixed.

**Pull request merged**